_posts/2016-09-22-approaching-tdd-outside-android-ii.md

```diff
@@ -0,0 +1,113 @@
+
+To show the statement lines we decided to use a <a href="https://developer.android.com/reference/android/support/v7/widget/RecyclerView.Adapter.html">Recyclerview</a>. Following the acceptance criteria each statement line has to be formatted as follows:
+
+<pre class="prettyprint"><code> 'Date' - 'Amount' - 'Running balance'</code></pre>
+
+It is worth mentioning that many people leave the UI out of the scope of the acceptance tests. In our case, we consider that in most mobile applications the UI is critical and therefore should generally be included. Whether you do it or not is your decision, however, we will show how it can be done. To assert the state of the view we are going to use Espresso, the official Android UI testing framework. Espresso does not offer a way to assert the state of each row in a RecyclerView out of the box, so we need to use the following snippet to be able to do so, <a href="https://gist.github.com/RomainPiel/ec10302a4687171a5e1a">RecyclerViewInteraction</a>. We consider that this piece of code is part of our testing framework. That is why it is already included as part of the initial steps.
+
+As we stated in the previous post, <b>Android is the responsible for instantiating Activities</b>. We can’t instantiate them manually. Bearing in mind that dependency injection through the constructor is not an option here, we had to come up with an alternative mechanism to instantiate custom dependencies for tests. For simplicity’s sake, we are not going to use any dependency injection framework. Instead, we are going to use the <a name="servicelocator" href="https://en.wikipedia.org/wiki/Service_locator_pattern">Service locator</a> pattern. Using this pattern and a static method <i>setInstance</i> we will be able to provide custom dependencies to the activity without using constructor dependency injection.
+
+Before asserting that the UI is showing the expected information, we need to define the layout that it is going to use to show it. In Android every view has an ID. We will need those IDs to reference the views in the acceptance test to assert that it is showing the correct information. 
+
+Having said that, let’s focus on the important stuff.
+
+Following the acceptance criteria, <a href="/2016/09/21/approaching-tdd-outside-android//#acceptance-criteria">defined in the previous post</a>, we have to make two deposits and one withdrawal. Notice that withdrawals are transactions with negative amounts.
+
+<b>As in Outside-In design happen in the red phase</b>, we have to design the skeleton of the classes that we know are needed at this point. We do not need to know and define the whole tree of collaborators, only the ones that we know that are needed. In contrast with classicist TDD, where everything emerges from the tests, in outside-in we need to do some design up front.
```

Comment:
maybe link to Sandro's post about that: https://codurance.com/2015/05/12/does-tdd-lead-to-good-design/